### ISSUE TITLE
dict not taking lists of lists

### ISSUE THREAD

**Issue by Dakkaron** [python/typeshed#714]
The `dict()` function can create a dict from lists of tuples, which works fine in mypy:

    dict( [ ("key", "value") ] )

But it can also create dicts from lists of lists, which mypy does not like:

    dict( [ ["key", "value"] ] )
    # List comprehension has incompatible type List[List[str]]

Also casting the inner list to tuple is blocked by mypy:

    dict( [ tuple(["key", "value"]) ] )
    # List comprehension has incompatible type List[Tuple[str, ...]]

This limits using list comprehension for the keys. In my current case I have a string containing key and value pairs:

    """"Key1     : Value1
    Key2     : Value2
    LongKey2 : Value""""

It's the output from some other program that is not under my control. As you see, the keys are padded by spaces to align the colon. So I am using this line, which works as expected:

     dict([ [y.strip() for y in x.split(":")] for x in res.splitlines() ])

But mypy does not like it, because dict only takes `List[Tuple[str, str]]` and not `List[List[str]]` or `List[Tuple[str,...]]`

**Comment by JukkaL** [comment]
I remember a similar issue being reported before. I don't think that we can express this very precisely, but perhaps we can do better than what we have right now.

Currently the cleanest option probably is to write this as a loop (not tested):

```
d = {}
for x in res.splitlines():
    a, b = [y.strip() for y in x.split(":")]
    d[a] = b
```

One option would be to have `dict` accept `Iterable[List[Union[_KT, _VT]]`, though this wouldn't be quite safe. I'm not sure whether mypy can infer the right key and value types based on that.

**Comment by gvanrossum** [comment]
I think mypy found a bug here though. The original code is not safe, it
will crash when the data has extra or missing colons.

**Comment by JelleZijlstra** [comment]
Duplicate of #256 (which I just closed; let's move the discussion there if we want to reopen this issue).
